src/Microsoft.Identity.Client/Azure/ManagedIdentity.cs

```diff
@@ -0,0 +1,370 @@
+using System.Net;
+using System.Net.Http;
+using System.Net.NetworkInformation;
+using System.Threading.Tasks;
+using System.Runtime.Serialization;
+using Microsoft.Identity.Client.Utils;
+using System.Text;
+
+namespace Microsoft.Identity.Client.Azure
+{
+#if !ANDROID_BUILDTIME && !iOS_BUILDTIME && !WINDOWS_APP_BUILDTIME && !MAC_BUILDTIME // Hide confidential client on mobile platforms
+
+    /// <summary>
+    ///     ManagedIdentityProbe will look in environment variable to determine if the managed identity provider is
+    ///     available. If the managed identity provider is available, the probe will allow for building a managed identity
+    ///     credential provider to fetch AAD tokens using the IMDS endpoint.
+    /// </summary>
+    public class ManagedIdentityProbe : IProbe
+    {
+
+        private readonly HttpClient _httpClient;
+        private readonly IManagedIdentityConfiguration _config;
+        private readonly string _overrideClientId;
+
+        internal ManagedIdentityProbe(HttpClient httpClient, IManagedIdentityConfiguration config = null, string overrideClientId = null)
+        {
+            _httpClient = httpClient;
+            _config = config ?? new DefaultManagedIdentityConfiguration();
+            _overrideClientId = overrideClientId;
+        }
+
+        /// <summary>
+        ///     Create a Managed Identity probe with a specified client identity
+        /// </summary>
+        /// <param name="config">option configuration structure -- if not supplied, a default environmental configuration is used.</param>
+        /// <param name="overrideClientId">override the client identity found in the config for use when querying the Azure IMDS endpoint</param>
+        public ManagedIdentityProbe(IManagedIdentityConfiguration config = null, string overrideClientId = null)
+            : this(null, config, overrideClientId) { }
+
+        /// <summary>
+        ///     Check if the probe is available for use in the current environment
+        /// </summary>
+        /// <returns>True if a credential provider can be built</returns>
+        public async Task<bool> AvailableAsync()
+        {
+            // check App Service MSI
+            if (IsAppService())
+            {
+                return true;
+            }
+
+#if net45  || NETSTANDARD2_0 
+
+            // Check if there is no service listening on VM IP
+            // 
+            // This is a performance optimization to avoid retrying requests to the Managed Identity service
+
+            if(!IsVMManagedIdentityListening())
+            {
+                return false;
+            }
+#endif
+
+            try
+            {
+                // if service is listening on VM IP check if a token can be acquired
+                var provider = BuildProvider(maxRetries: 3, httpClient: _httpClient);
+                var token = await provider.GetTokenAsync(new List<string> { @"https://management.azure.com//.default" }).ConfigureAwait(false);
```

Comment:
Does this mean that ARM needs to be added as an authorized APIs for Managed identities to can get a token our VM/service?